core/Segment.php

```diff
@@ -180,17 +182,45 @@ public function getSegmentExpression()
         return $this->segmentExpression;
     }
 
+    /**
+     * @throws Exception
+     */
     private function getAvailableSegments()
     {
+        // start lazy cache
+        $cache = PiwikCache::getLazyCache();
```

Comment:
I know there's no needs review label yet but wanted to mention using the lazy cache may be quite complicated as you will need to invalidate this cache whenever eg a custom report/form/goal/dimension... is added or changed or removed. Just some quick feedback. Maybe it's good enough to use a transient cache (haven't looked in detail)